src/App.tsx

```diff
@@ -36,35 +50,63 @@ function App(props: Props) {
       state.signer &&
       state.address &&
       roundId &&
-      (previousAddress === null || previousSigner === null)
+      (!previousAddress || !previousSigner)
     ) {
       const address = state.address;
       const message = `UMA Protocol one time key for round: ${roundId}`;
+      const hexMessage = web3.utils.toHex(message);
       const keyExists = signingKeys[address];
-      if (!keyExists || keyExists.roundMessage !== message) {
-        state.signer
-          .signMessage(message)
-          .then((msg) => {
-            const key = {} as {
-              publicKey: string;
-              privateKey: string;
-              roundMessage: string;
-            };
-
-            const privateKey = derivePrivateKey(msg);
-            const publicKey = recoverPublicKey(privateKey);
-            key.privateKey = privateKey;
-            key.publicKey = publicKey;
-            key.roundMessage = message;
-
-            setSigningKeys((prevKeys) => {
-              return { ...prevKeys, [address]: key };
+      let keysInStorage =
+        localStorage.getItem(SIGNING_KEYS_STORAGE_KEY) || ({} as StorageKeys);
```

Comment:
its a bit wierd to mix string and json here, makes this variable a double type, why not use "{}" as the fallback so yo u can treat it as an empty object if undefined

Reply:
Yeah that's fair.

Reply:
I actually get a type error if I try to swap it over... rather annoyingly. TypeScript is sometimes really frustrating to work with

Reply:
hmm ok i just tried this and it seemed to work:

```js
// parsing the json at the same time we fetch from storage
const keysInStorage = JSON.parse(localStorage.getItem(SIGNING_KEYS_STORAGE_KEY) || '{}') as StorageKeys;
// using array style path notation, saves some chars 
const keyExistsInStorage = has(keysInStorage, [hexMessage,address]);
```